wasmbinding/queries.go

```diff
@@ -95,19 +95,49 @@ func (qp QueryPlugin) EstimateSwap(ctx sdk.Context, estimateSwap *bindings.Estim
 	if err := sdk.ValidateDenom(estimateSwap.First.DenomOut); err != nil {
 		return nil, sdkerrors.Wrap(err, "gamm estimate swap denom out")
 	}
-	senderAddr, err := sdk.AccAddressFromBech32(estimateSwap.Sender)
-	if err != nil {
-		return nil, sdkerrors.Wrap(err, "gamm estimate swap sender address")
-	}
 
 	if estimateSwap.Amount == (bindings.SwapAmount{}) {
 		return nil, wasmvmtypes.InvalidRequest{Err: "gamm estimate swap empty swap"}
 	}
 
-	estimate, err := PerformSwap(qp.gammKeeper, ctx, senderAddr, estimateSwap.ToSwapMsg())
+	subCtx, _ := ctx.CacheContext()
+	estimate, err := EstimatePerformSwap(qp.gammKeeper, subCtx, estimateSwap.ToSwapMsg())
 	return estimate, err
 }
 
+// EstimatePerformSwap can be used to either estimate a exact amount in swap, or an exact amount out swap.
+// This function requires requires either swap.Amount.ExactIn or swap.Amount.ExactOut to be non nil, or else it will error.
+func EstimatePerformSwap(keeper *gammkeeper.Keeper, ctx sdk.Context, swap *bindings.SwapMsg) (*bindings.SwapAmount, error) {
+	if swap == nil {
+		return nil, wasmvmtypes.InvalidRequest{Err: "gamm perform swap null swap"}
+	}
+	if swap.Amount.ExactIn != nil {
```

Comment:
Why do you think that's the case? AFAIU, i think `swap.Amount.ExactIn != nil && swap.Amount.ExactOut == nil` can be a possible scenario, for example, when a pool is 1:1 ratio(10 uosmo, 10 uatom), you could be expecting 1 in 1out or similar

Reply:
Right, not only CAN it be a possible scenario, it should be the only scenario for the one sided argument. What I am trying to prevent here is if someone provides a SwapMsg with both a swap.Amount.ExactIn and swap.Amount.ExactOut, instead of erroring they would go through the first for loop. EstimatePerformSwap should only return a valid response if one side is given, not both sides

Reply:
AFAIU, we should never call an estimate method with both the in and the out. We either give it the in to calculate the out or we give it the out to calculate the in

Reply:
I still feel this is the case